rust/lance-table/src/format/manifest.rs

```diff
@@ -221,31 +222,46 @@ impl Manifest {
 
     /// Check the current fragment list and update the high water mark
     pub fn update_max_fragment_id(&mut self) {
+        // If there are no fragments, don't update max_fragment_id
+        if self.fragments.is_empty() {
+            return;
+        }
+
         let max_fragment_id = self
             .fragments
             .iter()
             .map(|f| f.id)
             .max()
-            .unwrap_or_default()
+            .unwrap() // Safe because we checked fragments is not empty
             .try_into()
             .unwrap();
 
-        if max_fragment_id > self.max_fragment_id {
-            self.max_fragment_id = max_fragment_id;
+        match self.max_fragment_id {
+            None => {
+                // First time being set
+                self.max_fragment_id = Some(max_fragment_id);
+            }
+            Some(current_max) => {
+                // Only update if the computed max is greater than current
+                // This preserves the high water mark even when fragments are deleted
+                if max_fragment_id > current_max {
+                    self.max_fragment_id = Some(max_fragment_id);
+                }
+            }
         }
     }
 
     /// Return the max fragment id.
     /// Note this does not support recycling of fragment ids.
     ///
-    /// This will return None if there are no fragments.
+    /// This will return None if there are no fragments and max_fragment_id was never set.
     pub fn max_fragment_id(&self) -> Option<u64> {
-        if self.max_fragment_id == 0 {
-            // It might not have been updated, so the best we can do is recompute
-            // it from the fragment list.
-            self.fragments.iter().map(|f| f.id).max()
+        if let Some(max_id) = self.max_fragment_id {
+            // Return the stored high water mark
+            Some(max_id.into())
         } else {
-            Some(self.max_fragment_id.into())
+            // Not yet set, compute from fragment list
```

Comment:
In what scenario do we end up with fragments but no max_fragment_id?

Reply:
This should only return `None` if an empty dataset was written with Lance v0.4.9 or earlier. Added a test for that. If there is a non-empty fragment list, then it should return `Some()`.